sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/PartitionMetadataDao.java

```diff
@@ -0,0 +1,470 @@
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.dao.PartitionMetadataAdminDao.COLUMN_WATERMARK;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TransactionContext;
+import com.google.cloud.spanner.TransactionRunner;
+import com.google.cloud.spanner.Value;
+import io.opencensus.common.Scope;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.PartitionMetadata;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.PartitionMetadata.State;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/** Data access object for the Connector metadata tables. */
+public class PartitionMetadataDao {
+  private static final Tracer TRACER = Tracing.getTracer();
+
+  private final String metadataTableName;
+  private final DatabaseClient databaseClient;
+
+  /**
+   * Constructs a partition metadata dao object given the generated name of the tables.
+   *
+   * @param metadataTableName the name of the partition metadata table
+   * @param databaseClient the {@link DatabaseClient} to perform queries
+   */
+  PartitionMetadataDao(String metadataTableName, DatabaseClient databaseClient) {
+    this.metadataTableName = metadataTableName;
+    this.databaseClient = databaseClient;
+  }
+
+  /**
+   * Fetches the partition metadata row data for the given partition token.
+   *
+   * @param partitionToken the partition unique identifier
+   * @return the partition metadata for the given token if it exists as a struct. Otherwise, it
+   *     returns null.
+   */
+  public @Nullable Struct getPartition(String partitionToken) {
+    try (Scope scope = TRACER.spanBuilder("getPartition").setRecordEvents(true).startScopedSpan()) {
+      TRACER
+          .getCurrentSpan()
+          .putAttribute(
+              PARTITION_ID_ATTRIBUTE_LABEL, AttributeValue.stringAttributeValue(partitionToken));
+      try (ResultSet resultSet =
+          databaseClient
+              .singleUse()
+              .executeQuery(
+                  Statement.newBuilder(
+                          "SELECT * FROM "
+                              + metadataTableName
+                              + " WHERE "
+                              + COLUMN_PARTITION_TOKEN
+                              + " = @partition")
+                      .bind("partition")
+                      .to(partitionToken)
+                      .build())) {
+        if (resultSet.next()) {
+          return resultSet.getCurrentRowAsStruct();
+        }
+        return null;
```

Comment:
How do we handle whenever the partition is missing? I suppose this should not happen right?

Reply:
It should not happen. If this happens we throw an `IllegalStateException` on the caller.